pagopa/io-app
# fix(IDPay): [IODPAY-179,IODPAY-180] Add support for gesture navigation in IDPay screens

## Short description
This PR adds the support for "swipe back" gesture to navigate through IDPay screens.

## List of changes proposed in this pull request
- [ts/features/idpay/initiative/configuration/navigation/navigator.tsx](https://github.com/pagopa/io-app/pull/4496/files#diff-0dc7659813959cbc98d6bc40b91064d21bd75497ce8e4ee693f7edc1364522d9)
  - Enabled gesture navigation in `IDPayConfigurationNavigator`
 - [ts/features/idpay/initiative/details/navigation/index.tsx](https://github.com/pagopa/io-app/pull/4496/files#diff-4a7124c38069898da2f4f3be7be79f02e4e9dda5586f5653d8cd4a2c0e88c0a4)
   - Enabled gesture navigation in `IDpayDetailsNavigator`
- [ts/features/idpay/onboarding/navigation/navigator.tsx](https://github.com/pagopa/io-app/pull/4496/files#diff-01b5ae5ce43bd13eb4995f7b94f39e84f2b77491d85513f111a2582d06e06115)
  - Enabled gesture navigation in `IDPayOnboardingNavigator`
- [ts/navigation/AppStackNavigator.tsx](https://github.com/pagopa/io-app/pull/4496/files#diff-8ffd33dd28dd9183ce57e6c092f399f95ee4843340c184fe84f0659ed9a1b66a)
  - Enabled gesture navigation for IDPay navigators/screens
- [ts/hooks/useNavigationSwipeBackListener.tsx](https://github.com/pagopa/io-app/pull/4496/files#diff-8a18070c46b39fbc529f59f2aa7470cc7515377c28243bfa0a0ace8ab7dfc26a):
  - Added `useNavigationSwipeBackListener`, an utility hook that allows to add a listener for the swipe back gesture
 - [ts/features/idpay/initiative/configuration/xstate/events.ts](https://github.com/pagopa/io-app/pull/4496/files#diff-22cbb26b3f180921c8f7d1d86041c6270e1999f2f26163896a19ee3d80ad5c9b)
   - Replaced `E_BACK` event with the global one `E_BACK` which supports `skipNavigation` property
 - [ts/features/idpay/onboarding/xstate/machine.ts](https://github.com/pagopa/io-app/pull/4496/files#diff-901fe4cd9014838a5f330338e00a7407e2163892d29ca5b564b2e9a98c61005d)
   - Moved `Events` in dedicated file `events.ts`
 - [ts/features/idpay/onboarding/xstate/events.ts](https://github.com/pagopa/io-app/pull/4496/files#diff-db9ceb1e0e2403ec9e630d9d576d9f3c8b920b4673c89f09fd4eced4643a65da)
   - Replaced `E_BACK` event with the global one `E_BACK` which supports `skipNavigation` property
 - [ts/features/idpay/onboarding/xstate/actions.ts](https://github.com/pagopa/io-app/pull/4496/files#diff-d338097022fceafc817f7828f772391496fdc78a90f24646044c665e8ba2bb67)
   - Added `skipNavigation` check in navigation actions
 - [ts/features/idpay/initiative/configuration/xstate/actions.ts](https://github.com/pagopa/io-app/pull/4496/files#diff-1af1026b4e387e27a8b35f3923bfeb5400e10f3b01e86a831ad5df8ebee9d0d7)
   - Added `skipNavigation` check in navigation actions
 - Added `useNavigationSwipeBackListener` to screens where the `BACK` event needed to be sent to the machine on gesture navigation
 - Added tests

## How to test
- Start the local dev server
- In the app, navigate to the IDPay Onboarding playground in the profile section
- Start an onboarding process
- During the onboarding steps try to navigate back using a swipe gesture
- Check that the process still works after some navigation actions;
